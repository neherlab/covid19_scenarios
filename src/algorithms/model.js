const msPerDay = 1000 * 60 * 60 * 24

const monthToDay = m => {
  return m * 30 + 15
}

const jan2020 = new Date('2020-01-01')

/**
 *
 * @param {number} time - point in time, until which simulation runs, as epoch time
 * @param {number} avgInfectionRate
 * @param {number} peakMonth - counting number in range of 0-11
 * @param {number} seasonalForcing -  seasonal variation in transmission. Usually a decimal number, e. g. 0.2
 * @returns {number}
 */
export function infectionRate(time, avgInfectionRate, peakMonth, seasonalForcing) {
  // this is super hacky
  const phase = ((time - jan2020) / msPerDay / 365 - monthToDay(peakMonth) / 365) * 2 * Math.PI
  return avgInfectionRate * (1 + seasonalForcing * Math.cos(phase))
}

export function getPopulationParams(params, severity, ageCounts, containment) {
  const pop = { ...params }

  pop.timeDeltaDays = 0.25
  pop.timeDelta = msPerDay * pop.timeDeltaDays
  pop.numberStochasticRuns = params.numberStochasticRuns

  // Compute age-stratified parameters
  const total = severity.map(d => d.ageGroup).reduce((a,b)=>a+ageCounts[b], 0);
  // TODO: Make this a form-adjustable factor
  pop.ageDistribution = {}
  pop.infectionSeverityRatio = {}
  pop.infectionFatality = {}
  pop.infectionCritical = {}
  pop.recoveryRate = {}
  pop.hospitalizedRate = {}
  pop.dischargeRate = {}
  pop.criticalRate = {}
  pop.deathRate = {}
  pop.overflowDeathRate = {}
  pop.stabilizationRate = {}
  pop.isolatedFrac = {}
  pop.importsPerDay = {}
  pop.importsPerDay.total = params.importsPerDay

  let hospitalizedFrac = 0
  let criticalFracHospitalized = 0
  let fatalFracCritical = 0
  let avgIsolatedFrac = 0
  severity.forEach(d => {
    const freq = (1.0 * ageCounts[d.ageGroup]) / total
    pop.ageDistribution[d.ageGroup] = freq
    pop.infectionSeverityRatio[d.ageGroup] = (d.severe / 100) * (d.confirmed / 100)
    pop.infectionCritical[d.ageGroup] = pop.infectionSeverityRatio[d.ageGroup] * (d.critical / 100)
    pop.infectionFatality[d.ageGroup] = pop.infectionCritical[d.ageGroup] * (d.fatal / 100)

    const dHospital = pop.infectionSeverityRatio[d.ageGroup]
    const dCritical = d.critical / 100
    const dFatal = d.fatal / 100

    hospitalizedFrac += freq * dHospital
    criticalFracHospitalized += freq * dCritical
    fatalFracCritical += freq * dFatal
    avgIsolatedFrac += (freq * d.isolated) / 100

    // Age specific rates
    pop.isolatedFrac[d.ageGroup] = d.isolated / 100
    pop.recoveryRate[d.ageGroup] = (1 - dHospital) / pop.infectiousPeriod
    pop.hospitalizedRate[d.ageGroup] = dHospital / pop.infectiousPeriod
    pop.dischargeRate[d.ageGroup] = (1 - dCritical) / pop.lengthHospitalStay
    pop.criticalRate[d.ageGroup] = dCritical / pop.lengthHospitalStay
    pop.stabilizationRate[d.ageGroup] = (1 - dFatal) / pop.lengthICUStay
    pop.deathRate[d.ageGroup] = dFatal / pop.lengthICUStay
    pop.overflowDeathRate[d.ageGroup] = pop.overflowSeverity*pop.deathRate[d.ageGroup];
  })

  // Get import rates per age class (assume flat)
  const L = Object.keys(pop.recoveryRate).length
  Object.keys(pop.recoveryRate).forEach(k => {
    pop.importsPerDay[k] = params.importsPerDay / L
  })

  // Population average rates
  pop.recoveryRate.total = (1 - hospitalizedFrac) / pop.infectiousPeriod
  pop.hospitalizedRate.total = hospitalizedFrac / pop.infectiousPeriod
  pop.dischargeRate.total = (1 - criticalFracHospitalized) / pop.lengthHospitalStay
  pop.criticalRate.total = criticalFracHospitalized / pop.lengthHospitalStay
  pop.deathRate.total = fatalFracCritical / pop.lengthICUStay
  pop.stabilizationRate.total = (1 - fatalFracCritical) / pop.lengthICUStay
  pop.overflowDeathRate.total = pop.overflowSeverity*fatalFracCritical / pop.lengthICUStay
  pop.isolatedFrac.total = avgIsolatedFrac

  // Infectivity dynamics
  const avgInfectionRate = pop.r0 / pop.infectiousPeriod
  pop.infectionRate = time =>
    containment(time) * infectionRate(time, avgInfectionRate, pop.peakMonth, pop.seasonalForcing)

  return pop
}

export function initializePopulation(N, numCases, t0, ages) {
  // FIXME: Why it can be `undefined`? Can it also be `null`?
  if (ages === undefined) {
    const put = x => {
      return { total: x }
    }
    return {
      time: t0,
      susceptible: put(N - numCases),
      exposed: put(0),
      infectious: put(numCases),
      hospitalized: put(0),
      critical: put(0),
      discharged: put(0),
      recovered: put(0),
      intensive: put(0),
      dead: put(0),
    }
  }
  const Z = Object.values(ages).reduce((a, b) => a + b)
  const pop = {
    time: t0,
    susceptible: {},
    exposed: {},
    infectious: {},
    hospitalized: {},
    critical: {},
    overflow: {},
    discharged: {},
    intensive: {},
    recovered: {},
    dead: {},
  }
  // TODO: Ensure the sum is equal to N!
  Object.keys(ages).forEach((k, i) => {
    const n = Math.round((ages[k] / Z) * N)
    pop.susceptible[k] = n
    pop.exposed[k] = 0
    pop.infectious[k] = 0
    pop.hospitalized[k] = 0
    pop.critical[k] = 0
    pop.overflow[k] = 0
    pop.discharged[k] = 0
    pop.recovered[k] = 0
    pop.intensive[k] = 0
    pop.dead[k] = 0
    if (i === Math.round(Object.keys(ages).length / 2)) {
      pop.susceptible[k] -= numCases
      pop.infectious[k] = 0.3 * numCases
      pop.exposed[k] = 0.7 * numCases
    }
  })

  return pop
}

// NOTE: Assumes all subfields corresponding to populations have the same set of keys
export function evolve(pop, P, sample) {
  const sum = (dict) => { return Object.values(dict).reduce((a, b) => a + b, 0); }
  const fracInfected = sum(pop.infectious) / P.populationServed

  const newTime = pop.time + P.timeDelta
  const newPop = {
    time: newTime,
    susceptible: {},
    exposed: {},
    infectious: {},
    recovered: {},
    hospitalized: {},
    critical: {},
    overflow: {},
    discharged: {},
    intensive: {},
    dead: {},
  }

  const push = (sub, age, delta) => {
    newPop[sub][age] = pop[sub][age] + delta
  }

  const newCases        = {};
  const newInfectious   = {};
  const newRecovered    = {};
  const newHospitalized = {};
  const newDischarged   = {};
  const newCritical     = {};
  const newStabilized   = {};
  const newICUDead      = {};
  const newOverflowStabilized = {};
  const newOverflowDead = {};

  // Compute all fluxes (apart from overflow states) barring no hospital bed constraints
  const Keys = Object.keys(pop.infectious).sort();
  Keys.forEach(age => {
    newCases[age] =
      sample(P.importsPerDay[age] * P.timeDeltaDays) +
      sample(
        (1 - P.isolatedFrac[age]) * P.infectionRate(newTime) * pop.susceptible[age] * fracInfected * P.timeDeltaDays,
      )
    newInfectious[age]         = Math.min(pop.exposed[age], sample(pop.exposed[age] * P.timeDeltaDays / P.incubationTime))
    newRecovered[age]          = Math.min(pop.infectious[age], sample(pop.infectious[age] * P.timeDeltaDays * P.recoveryRate[age]))
    newHospitalized[age]       = Math.min(pop.infectious[age] - newRecovered[age], sample(pop.infectious[age] * P.timeDeltaDays * P.hospitalizedRate[age]))
    newDischarged[age]         = Math.min(pop.hospitalized[age], sample(pop.hospitalized[age] * P.timeDeltaDays * P.dischargeRate[age]))
    newCritical[age]           = Math.min(pop.hospitalized[age] - newDischarged[age], sample(pop.hospitalized[age] * P.timeDeltaDays * P.criticalRate[age]))
    newStabilized[age]         = Math.min(pop.critical[age], sample(pop.critical[age] * P.timeDeltaDays * P.stabilizationRate[age]))
    newICUDead[age]            = Math.min(pop.critical[age] - newStabilized[age], sample(pop.critical[age] * P.timeDeltaDays * P.deathRate[age]))
    newOverflowStabilized[age] = Math.min(pop.overflow[age], sample(pop.overflow[age] * P.timeDeltaDays * P.stabilizationRate[age]))
    newOverflowDead[age]       = Math.min(pop.overflow[age] - newOverflowStabilized[age], sample(pop.overflow[age] * P.timeDeltaDays * P.overflowDeathRate[age]))

    push('susceptible', age, -newCases[age])
    push('exposed', age, newCases[age] - newInfectious[age])
    push('infectious', age, newInfectious[age] - newRecovered[age] - newHospitalized[age])
    push('hospitalized', age, newHospitalized[age] + newStabilized[age] + newOverflowStabilized[age] - newDischarged[age] - newCritical[age])
    // Cumulative categories
    push('recovered', age, newRecovered[age] + newDischarged[age])
    push('intensive', age, newCritical[age])
    push('discharged', age, newDischarged[age])
    push('dead', age, newICUDead[age] + newOverflowDead[age])
  })

  // Move hospitalized patients according to constrained resources
  let freeICUBeds = P.ICUBeds - (sum(pop.critical) - sum(newStabilized) - sum(newICUDead));
  Keys.forEach(age => {
      if (freeICUBeds > newCritical[age]) {
          freeICUBeds -= newCritical[age];
          push('critical', age, newCritical[age] - newStabilized[age] - newICUDead[age])
          push('overflow', age,  - newOverflowDead[age] - newOverflowStabilized[age])
      } else if (freeICUBeds > 0) {
          let newOverflow = newCritical[age] - freeICUBeds;
          push('critical', age, freeICUBeds - newStabilized[age] - newICUDead[age])
          push('overflow', age, newOverflow - newOverflowDead[age] - newOverflowStabilized[age])
          freeICUBeds = 0;
      } else {
          push('critical', age, -newStabilized[age] - newICUDead[age])
          push('overflow', age, newCritical[age] - newOverflowDead[age] - newOverflowStabilized[age])
      }
  });

  // If any overflow patients are left AND there are free beds, move them back.
  // Again, move w/ lower age as priority.
  for (let i = 0; freeICUBeds > 0 && i < Keys.length; i++) {
      const age = Keys[i];
      if (newPop.overflow[age] < freeICUBeds) {
          newPop.critical[age] += newPop.overflow[age];
          freeICUBeds -= newPop.overflow[age];
          newPop.overflow[age] = 0;
      } else {
          newPop.critical[age] += freeICUBeds;
          newPop.overflow[age] -= freeICUBeds;
          freeICUBeds = 0;
      }
  }

  // NOTE: For debug purposes only.
  /*
  const popSum = sum(newPop.susceptible) + sum(newPop.exposed) + sum(newPop.infectious) + sum(newPop.recovered) + sum(newPop.hospitalized) + sum(newPop.critical) + sum(newPop.overflow) + sum(newPop.dead);
  console.log(math.abs(popSum - P.populationServed));
  */

  return newPop
}

export function collectTotals(trajectory) {
  // FIXME: parameter reassign
  trajectory.forEach(d => {
    Object.keys(d).forEach(k => {
      if (k === 'time' || 'total' in d[k]) {
        return
      }
      d[k].total = Object.values(d[k]).reduce((a, b) => a + b)
    })
  })

  return trajectory
}

export function exportSimulation(trajectory) {
  // Store parameter values

  // Down sample trajectory to once a day.
  // TODO: Make the down sampling interval a parameter

  const header = Object.keys(trajectory[0])
  const csv = [header.join('\t')]

  const pop = {}
  trajectory.forEach(d => {
    const t = new Date(d.time).toISOString().slice(0, 10)
    if (t in pop) {
      return
    } // skip if date is already in table
    pop[t] = true
    let buf = ''
    header.forEach(k => {
      if (k === 'time') {
        buf += `${t}`
      } else {
        buf += `\t${Math.round(d[k].total)}`
      }
    })
    csv.push(buf)
  })

  return csv.join('\n')
}
