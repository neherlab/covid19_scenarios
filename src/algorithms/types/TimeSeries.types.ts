export interface TimePoint {
  t: Date
  y: number
}

export type TimeSeries = TimePoint[]
