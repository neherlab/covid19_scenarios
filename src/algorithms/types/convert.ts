import { omit } from 'lodash'

import { uuidv4 } from '../../helpers/uuid'

import {
  MitigationAgeSpecificDatum,
  MitigationAgeSpecificDatumExternal,
  MitigationInterval,
  MitigationIntervalExternal,
  ScenarioDatum,
  ScenarioDatumExternal,
} from './Param.types'

export function mitigtaionIntervalAgeSpecificToInternal(
  mitigationAgeSpecificData: MitigationAgeSpecificDatumExternal[],
): MitigationAgeSpecificDatum[] {
  return mitigationAgeSpecificData.map((datum) => ({ ...datum, id: uuidv4() } as MitigationAgeSpecificDatum))
}

export function mitigtaionIntervalToInternal(interval: MitigationIntervalExternal): MitigationInterval {
  return {
    ...interval,
    id: uuidv4(),
    mitigationAgeSpecificData: mitigtaionIntervalAgeSpecificToInternal(interval.mitigationAgeSpecificData),
  }
}

export function toInternal(scenario: ScenarioDatumExternal): ScenarioDatum {
  const { mitigationIntervals } = scenario.mitigation
  return {
    ...scenario,
    mitigation: {
      ...scenario.mitigation,
      mitigationIntervals: mitigationIntervals.map(mitigtaionIntervalToInternal),
    },
  }
}

export function mitigtaionIntervalAgeSpecificToExternal(
  mitigationAgeSpecificData: MitigationAgeSpecificDatum[],
): MitigationAgeSpecificDatumExternal[] {
  return mitigationAgeSpecificData.map((datum) => omit(datum, 'id'))
}

export function mitigationIntervalToExternal(interval: MitigationInterval): MitigationIntervalExternal {
  const intervalWithoutId = omit(interval, 'id')
  return {
    ...intervalWithoutId,
    mitigationAgeSpecificData: mitigtaionIntervalAgeSpecificToExternal(intervalWithoutId.mitigationAgeSpecificData),
  }
}

export function toExternal(scenario: ScenarioDatum): ScenarioDatumExternal {
  const { mitigationIntervals } = scenario.mitigation
  return {
    ...scenario,
    mitigation: {
      ...scenario.mitigation,
      mitigationIntervals: mitigationIntervals.map(mitigationIntervalToExternal),
    },
  }
}
