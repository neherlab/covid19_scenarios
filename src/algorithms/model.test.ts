import { infectionRate, getPopulationParams, initializePopulation, evolve } from './model'
import { interpolateTimeSeries } from './run'

// The input and output of getPopulationParams with the default settings
import getPopulationParamsInput from './__test_data__/getPopulationParams.input.default.json'

// The containment information with default settings
import { containment } from './__test_data__/containment.default.json'

// The input and output to initializePopulation with default settings
import initializePopulationInput from './__test_data__/initializePopulation.input.default.json'

// The output after the 5th generation using default settings
// This is a snapshot of the output that is being used as baseline
import { AllParamsFlat } from './types/Param.types'

const containmentWithDate = containment.map((value) => ({ y: value.y, t: new Date(value.t) }))
const containmentFunction = interpolateTimeSeries(containmentWithDate)

// The snapshot we load has the dates loaded as strings
// This just overwrites the simulationTimeRange field with the correct object types
const processParams = (params: typeof getPopulationParamsInput.params): AllParamsFlat => {
  return {
    ...params,
    simulationTimeRange: {
      tMin: new Date(params.simulationTimeRange.tMin),
      tMax: new Date(params.simulationTimeRange.tMax),
    },
  }
}

const identity = (x: number) => x

describe('model', () => {
  describe('infectionRate', () => {
    const dec2020 = new Date('2020-12-01').getTime()

    it('baseline', () => {
      expect(infectionRate(dec2020, 0.9, 3, 0.2)).toEqual(0.7768945041075702)
    })

    it('accounts for time correctly', () => {
      const jul2020 = new Date('2020-07-01').getTime()
      const feb2021 = new Date('2021-02-01').getTime()

      expect(infectionRate(jul2020, 0.4, 3, 0.2)).toEqual(0.4194279777676749)
      expect(infectionRate(feb2021, 1.5, 3, 0.2)).toEqual(1.5927050983124844)
    })

    it('accounts for avgInfectionRate correctly', () => {
      expect(infectionRate(dec2020, 0.4, 3, 0.2)).toEqual(0.3452864462700312)
      expect(infectionRate(dec2020, 1.5, 3, 0.2)).toEqual(1.294824173512617)
    })

    it('accounts for peakMonth correctly', () => {
      expect(infectionRate(dec2020, 0.9, 2, 0.2)).toEqual(0.8577915498773262)
      expect(infectionRate(dec2020, 0.9, 7, 0.2)).toEqual(0.842905568053961)
    })

    it('accounts for seasonalForcing correctly', () => {
      expect(infectionRate(dec2020, 0.9, 3, 0.1)).toEqual(0.8384472520537851)
      expect(infectionRate(dec2020, 0.9, 3, 0.4)).toEqual(0.6537890082151402)
    })
  })

  describe('getPopulationParams', () => {
    it('produces the expected output in the default scenario', () => {
      const inputParams = processParams(getPopulationParamsInput.params)
      const params = getPopulationParams(
        inputParams,
        getPopulationParamsInput.severity,
        getPopulationParamsInput.ageCounts,
        containmentFunction,
      )

      expect(params).toMatchSnapshot()
      containmentWithDate.forEach((o) => {
        expect(params.rate.infection(o.t)).toMatchSnapshot()
      })
    })
  })

  describe('initializePopulation', () => {
    it('produces the expected output in the default scenario', () => {
      const result = initializePopulation(
        initializePopulationInput.N,
        initializePopulationInput.numCases,
        initializePopulationInput.t0,
        initializePopulationInput.ages,
      )
      expect(result.current.susceptible).toMatchSnapshot()
    })
  })

  describe('evolve', () => {
    it('model evolution produces the expected output in the default scenario after running for 10 days', () => {
      const params = getPopulationParams(
        processParams(getPopulationParamsInput.params),
        getPopulationParamsInput.severity,
        getPopulationParamsInput.ageCounts,
        containmentFunction,
      )

      const input = initializePopulation(
        initializePopulationInput.N,
        initializePopulationInput.numCases,
        initializePopulationInput.t0,
        initializePopulationInput.ages,
      )

      // run model for 0.25 days
      const result = evolve(input, params, input.time + 10, identity)
      // compare result with snapshot
      expect(result).toMatchSnapshot()
    })
  })
})
