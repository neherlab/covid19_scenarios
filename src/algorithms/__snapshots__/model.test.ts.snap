// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model evolve produces correct output for 10 days 1`] = `
Object {
  "cumulative": Object {
    "critical": Array [
      0.00022557140737396636,
      0.0012787144115711316,
      0.003183838401432405,
      0.008410398294746059,
      0.028652959760651776,
      0.9217587220622673,
      0.2633506865255802,
      0.4867356090602858,
      0.6507614033157809,
    ],
    "fatality": Array [
      0.000008661040932724528,
      0.000049080151206733994,
      0.00012220146424743398,
      0.0003226874447001694,
      0.0010989560654547312,
      0.08713414347985174,
      0.013457900108679664,
      0.031087669615630267,
      0.04154501211964636,
    ],
    "hospitalized": Array [
      0.01128474462586532,
      0.03194177264050325,
      0.0795324967157279,
      0.13987296237157062,
      0.3569081902448129,
      5.490424419403363,
      1.8693769961005962,
      2.685637040581435,
      2.9319219308119013,
    ],
    "recovered": Array [
      22.562490363844887,
      21.27408175073247,
      26.45995428747313,
      30.990666488314353,
      29.50005283586553,
      216.89182852291796,
      23.544729464978676,
      17.092383755597737,
      9.328206940603161,
    ],
  },
  "current": Object {
    "critical": Array [
      0.00019670127093155128,
      0.0011151139075486851,
      0.0027765001872742923,
      0.007334773479078828,
      0.02498977287580267,
      0.7039233633626382,
      0.22970593625388108,
      0.42456026982902517,
      0.5676713790764882,
    ],
    "exposed": Array [
      Array [
        26.1743878275881,
        21.364312704594642,
        17.24896816298584,
      ],
      Array [
        24.6946739655864,
        20.156703669053652,
        16.27412953650915,
      ],
      Array [
        30.74864512971563,
        25.097409075987724,
        20.26249856238672,
      ],
      Array [
        36.054888544429105,
        29.42788651283459,
        23.758262963182982,
      ],
      Array [
        34.49909010441671,
        28.158184104898776,
        22.73329965555768,
      ],
      Array [
        38.89489082056056,
        32.50751806405501,
        28.62787917565346,
      ],
      Array [
        28.90818542369758,
        23.595391809264324,
        19.04999980977432,
      ],
      Array [
        22.24422583241341,
        18.15686884778544,
        14.659769116067276,
      ],
      Array [
        13.591572788297185,
        11.09535360095183,
        8.959382963372056,
      ],
    ],
    "infectious": Array [
      20.207924848469567,
      19.06609534176138,
      23.737676169295295,
      27.832268682269223,
      26.63172808390407,
      52.66985936671702,
      22.317475107316945,
      17.175193286326213,
      10.498325000646103,
    ],
    "overflow": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "severe": Array [
      0.006793525573895686,
      0.019269148877607643,
      0.047979249451314536,
      0.08455657777756395,
      0.2162076222609483,
      1.9340907463740713,
      1.1371333133333865,
      1.63509002339643,
      1.7902643914483096,
    ],
    "susceptible": Array [
      19831.434917204635,
      18709.513882393418,
      23299.642938824036,
      27322.84381277027,
      26143.235348864157,
      29382.682875796883,
      21904.20392123527,
      16851.580821198968,
      10291.127677923485,
    ],
  },
  "time": 1583913600000,
}
`;

exports[`model getPopulationParams calculates infection rates for 2020-07-01T00:00:00.000Z 1`] = `
Object {
  "ICUBeds": 80,
  "ageDistribution": Object {
    "0-9": 0.10225114395521838,
    "10-19": 0.0964647560575147,
    "20-29": 0.12013546056661978,
    "30-39": 0.14087582334618426,
    "40-49": 0.13479387083572222,
    "50-59": 0.15259164409491252,
    "60-69": 0.11293803472872646,
    "70-79": 0.08688929452955889,
    "80+": 0.053059971885542775,
  },
  "frac": Object {
    "critical": Array [
      0.000025,
      0.00015000000000000001,
      0.00030000000000000003,
      0.0006749999999999999,
      0.0024000000000000002,
      0.00625,
      0.02625,
      0.063,
      0.1375,
    ],
    "fatal": Array [
      0.0000075,
      0.000045,
      0.00009,
      0.00020249999999999996,
      0.00072,
      0.0025000000000000005,
      0.0105,
      0.0315,
      0.06875,
    ],
    "isolated": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "severe": Array [
      0.0005,
      0.0015,
      0.003,
      0.0045,
      0.012,
      0.025,
      0.075,
      0.13999999999999999,
      0.25,
    ],
  },
  "hospitalBeds": 698,
  "importsPerDay": Array [
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
  ],
  "numberStochasticRuns": 0,
  "populationServed": 195000,
  "rate": Object {
    "critical": Array [
      0.0125,
      0.025,
      0.025,
      0.0375,
      0.05,
      0.0625,
      0.0875,
      0.1125,
      0.1375,
    ],
    "discharge": Array [
      0.2375,
      0.225,
      0.225,
      0.2125,
      0.2,
      0.1875,
      0.1625,
      0.1375,
      0.11249999999999999,
    ],
    "fatality": Array [
      0.02142857142857143,
      0.02142857142857143,
      0.02142857142857143,
      0.02142857142857143,
      0.02142857142857143,
      0.028571428571428574,
      0.028571428571428574,
      0.03571428571428571,
      0.03571428571428571,
    ],
    "infection": [Function],
    "latency": 0.2,
    "overflowFatality": Array [
      0.04285714285714286,
      0.04285714285714286,
      0.04285714285714286,
      0.04285714285714286,
      0.04285714285714286,
      0.05714285714285715,
      0.05714285714285715,
      0.07142857142857142,
      0.07142857142857142,
    ],
    "recovery": Array [
      0.33316666666666667,
      0.33283333333333337,
      0.3323333333333333,
      0.33183333333333337,
      0.3293333333333333,
      0.325,
      0.30833333333333335,
      0.2866666666666667,
      0.25,
    ],
    "severe": Array [
      0.00016666666666666666,
      0.0005,
      0.001,
      0.0014999999999999998,
      0.004,
      0.008333333333333333,
      0.024999999999999998,
      0.04666666666666666,
      0.08333333333333333,
    ],
    "stabilize": Array [
      0.049999999999999996,
      0.049999999999999996,
      0.049999999999999996,
      0.049999999999999996,
      0.049999999999999996,
      0.04285714285714286,
      0.04285714285714286,
      0.03571428571428571,
      0.03571428571428571,
    ],
  },
  "timeDelta": 0,
  "timeDeltaDays": 0,
}
`;

exports[`model getPopulationParams calculates infection rates for 2020-12-01T00:00:00.000Z 1`] = `
Object {
  "ICUBeds": 80,
  "ageDistribution": Object {
    "0-9": 0.10225114395521838,
    "10-19": 0.0964647560575147,
    "20-29": 0.12013546056661978,
    "30-39": 0.14087582334618426,
    "40-49": 0.13479387083572222,
    "50-59": 0.15259164409491252,
    "60-69": 0.11293803472872646,
    "70-79": 0.08688929452955889,
    "80+": 0.053059971885542775,
  },
  "frac": Object {
    "critical": Array [
      0.000025,
      0.00015000000000000001,
      0.00030000000000000003,
      0.0006749999999999999,
      0.0024000000000000002,
      0.00625,
      0.02625,
      0.063,
      0.1375,
    ],
    "fatal": Array [
      0.0000075,
      0.000045,
      0.00009,
      0.00020249999999999996,
      0.00072,
      0.0025000000000000005,
      0.0105,
      0.0315,
      0.06875,
    ],
    "isolated": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "severe": Array [
      0.0005,
      0.0015,
      0.003,
      0.0045,
      0.012,
      0.025,
      0.075,
      0.13999999999999999,
      0.25,
    ],
  },
  "hospitalBeds": 698,
  "importsPerDay": Array [
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
  ],
  "numberStochasticRuns": 0,
  "populationServed": 195000,
  "rate": Object {
    "critical": Array [
      0.0125,
      0.025,
      0.025,
      0.0375,
      0.05,
      0.0625,
      0.0875,
      0.1125,
      0.1375,
    ],
    "discharge": Array [
      0.2375,
      0.225,
      0.225,
      0.2125,
      0.2,
      0.1875,
      0.1625,
      0.1375,
      0.11249999999999999,
    ],
    "fatality": Array [
      0.02142857142857143,
      0.02142857142857143,
      0.02142857142857143,
      0.02142857142857143,
      0.02142857142857143,
      0.028571428571428574,
      0.028571428571428574,
      0.03571428571428571,
      0.03571428571428571,
    ],
    "infection": [Function],
    "latency": 0.2,
    "overflowFatality": Array [
      0.04285714285714286,
      0.04285714285714286,
      0.04285714285714286,
      0.04285714285714286,
      0.04285714285714286,
      0.05714285714285715,
      0.05714285714285715,
      0.07142857142857142,
      0.07142857142857142,
    ],
    "recovery": Array [
      0.33316666666666667,
      0.33283333333333337,
      0.3323333333333333,
      0.33183333333333337,
      0.3293333333333333,
      0.325,
      0.30833333333333335,
      0.2866666666666667,
      0.25,
    ],
    "severe": Array [
      0.00016666666666666666,
      0.0005,
      0.001,
      0.0014999999999999998,
      0.004,
      0.008333333333333333,
      0.024999999999999998,
      0.04666666666666666,
      0.08333333333333333,
    ],
    "stabilize": Array [
      0.049999999999999996,
      0.049999999999999996,
      0.049999999999999996,
      0.049999999999999996,
      0.049999999999999996,
      0.04285714285714286,
      0.04285714285714286,
      0.03571428571428571,
      0.03571428571428571,
    ],
  },
  "timeDelta": 0,
  "timeDeltaDays": 0,
}
`;

exports[`model getPopulationParams calculates infection rates for 2021-02-01T00:00:00.000Z 1`] = `
Object {
  "ICUBeds": 80,
  "ageDistribution": Object {
    "0-9": 0.10225114395521838,
    "10-19": 0.0964647560575147,
    "20-29": 0.12013546056661978,
    "30-39": 0.14087582334618426,
    "40-49": 0.13479387083572222,
    "50-59": 0.15259164409491252,
    "60-69": 0.11293803472872646,
    "70-79": 0.08688929452955889,
    "80+": 0.053059971885542775,
  },
  "frac": Object {
    "critical": Array [
      0.000025,
      0.00015000000000000001,
      0.00030000000000000003,
      0.0006749999999999999,
      0.0024000000000000002,
      0.00625,
      0.02625,
      0.063,
      0.1375,
    ],
    "fatal": Array [
      0.0000075,
      0.000045,
      0.00009,
      0.00020249999999999996,
      0.00072,
      0.0025000000000000005,
      0.0105,
      0.0315,
      0.06875,
    ],
    "isolated": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "severe": Array [
      0.0005,
      0.0015,
      0.003,
      0.0045,
      0.012,
      0.025,
      0.075,
      0.13999999999999999,
      0.25,
    ],
  },
  "hospitalBeds": 698,
  "importsPerDay": Array [
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
    0.011111111111111112,
  ],
  "numberStochasticRuns": 0,
  "populationServed": 195000,
  "rate": Object {
    "critical": Array [
      0.0125,
      0.025,
      0.025,
      0.0375,
      0.05,
      0.0625,
      0.0875,
      0.1125,
      0.1375,
    ],
    "discharge": Array [
      0.2375,
      0.225,
      0.225,
      0.2125,
      0.2,
      0.1875,
      0.1625,
      0.1375,
      0.11249999999999999,
    ],
    "fatality": Array [
      0.02142857142857143,
      0.02142857142857143,
      0.02142857142857143,
      0.02142857142857143,
      0.02142857142857143,
      0.028571428571428574,
      0.028571428571428574,
      0.03571428571428571,
      0.03571428571428571,
    ],
    "infection": [Function],
    "latency": 0.2,
    "overflowFatality": Array [
      0.04285714285714286,
      0.04285714285714286,
      0.04285714285714286,
      0.04285714285714286,
      0.04285714285714286,
      0.05714285714285715,
      0.05714285714285715,
      0.07142857142857142,
      0.07142857142857142,
    ],
    "recovery": Array [
      0.33316666666666667,
      0.33283333333333337,
      0.3323333333333333,
      0.33183333333333337,
      0.3293333333333333,
      0.325,
      0.30833333333333335,
      0.2866666666666667,
      0.25,
    ],
    "severe": Array [
      0.00016666666666666666,
      0.0005,
      0.001,
      0.0014999999999999998,
      0.004,
      0.008333333333333333,
      0.024999999999999998,
      0.04666666666666666,
      0.08333333333333333,
    ],
    "stabilize": Array [
      0.049999999999999996,
      0.049999999999999996,
      0.049999999999999996,
      0.049999999999999996,
      0.049999999999999996,
      0.04285714285714286,
      0.04285714285714286,
      0.03571428571428571,
      0.03571428571428571,
    ],
  },
  "timeDelta": 0,
  "timeDeltaDays": 0,
}
`;

exports[`model initializePopulation produces the expected output in the default scenario 1`] = `
Object {
  "critical": Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "exposed": Array [
    Array [
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
    ],
    Array [
      49.699999999999996,
      49.699999999999996,
      49.699999999999996,
    ],
    Array [
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
    ],
  ],
  "infectious": Array [
    0,
    0,
    0,
    0,
    0,
    63.9,
    0,
    0,
    0,
  ],
  "overflow": Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "severe": Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "susceptible": Array [
    19939,
    18811,
    23426,
    27471,
    26285,
    29542,
    22023,
    16943,
    10347,
  ],
}
`;
